### ISSUE TITLE
Status of Flash-Attention on TPUs

### ISSUE THREAD

**Issue by jon-chuang** [openxla/xla#2338]
Hello, [Flash Attention](https://arxiv.org/abs/2205.14135) is a method to produce tiled and fused kernels such that the tiled parameters can fit onto the device SRAM.

May I ask to what degree this technique has been applied to TPUs? 

I am aware that there is ongoing work on improving tiling via [gml_st](https://github.com/tensorflow/mlir-hlo/tree/master/gml_st), but it doesn't seem to be a blocker for this work, since we already seem to support rectangular tiling well in XLA (?). 

May I also know the status of integration with gml_st?

Resources
1. Triton [example implementation](https://github.com/openai/triton/blob/main/python/tutorials/06-fused-attention.py)
2. https://github.com/HazyResearch/flash-attention
3. https://github.com/lucidrains/flash-attention-jax

**Comment by jon-chuang** [comment]
CC: @frgossen

**Comment by blakehechtman** [comment]
TPUs have deployed an automatic version of Self Attention fusion before Flash Attention was ever published or popular in the community. The main limitation is that the automatic fusion algorithm avoids recomputing softmax(qk) when training a model. Also the divide operations is not reassociated with the v @ softmax(qk) to allow the softmax to be partially computed, so the SRAM footprint of the full sequence limits maximum sequence length that can automatically fuse on TPUs
